src/ice_nucleation.F90

```diff
@@ -0,0 +1,563 @@
+    allocate(H2O_masses(aero_state_n_part(aero_state)))
+    allocate(H2O_frac(aero_state_n_part(aero_state)))
+
+    total_masses = aero_state_masses(aero_state, aero_data)
+    H2O_masses = aero_state_masses(aero_state, aero_data, include=(/"H2O"/))
+    H2O_frac = H2O_masses / total_masses
+    do i_part = 1, aero_state_n_part(aero_state)
+       if (aero_state%apa%particle(i_part)%frozen) then
+          cycle
+       end if
+       if (H2O_frac(i_part) < 1e-2) then
+          cycle
+       end if
+       if (env_state_final%temp .le. &
+            aero_state%apa%particle(i_part)%imf_temperature) then
+          aero_state%apa%particle(i_part)%frozen = .TRUE.
+          aero_state%apa%particle(i_part)%den_ice = &
+               const%reference_ice_density
+          aero_state%apa%particle(i_part)%ice_shape_phi = 1d0
+       end if
+    end do
+
+    deallocate(total_masses)
+    deallocate(H2O_masses)
+    deallocate(H2O_frac)
+
+  end subroutine immersion_freezing_singular
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Simulation for time-dependent scheme (e.g., ABIFM, constant rate),
+  !> deciding whether to freeze for each particle. Run in each time step.   
+  !> This subroutine applys the binned-tau leaping algorithm for speeding up.
+  subroutine immersion_freezing_time_dependent(aero_state, aero_data, env_state_initial, &
+       env_state_final, del_t, immersion_freezing_scheme_type, freezing_rate)
+
+    implicit none
+    !> Aerosol state.
+    type(aero_state_t), intent(inout) :: aero_state
+    !> Aerosol data.
+    type(aero_data_t), intent(in) :: aero_data
+    !> Environment state at the start of the timestep.
+    type(env_state_t), intent(in) :: env_state_initial
+    !> Environment state at the end of the timestep. The rel_humid
+    !> value will be ignored and overwritten with a new value.
+    type(env_state_t), intent(inout) :: env_state_final
+    !> Total time to integrate.
+    real(kind=dp), intent(in) :: del_t
+
+    integer :: i_part, i_bin, i_class, n_bins, n_class
+    real(kind=dp) :: tmp
+    integer, intent(in) :: immersion_freezing_scheme_type
+    real(kind=dp) :: freezing_rate
+    real(kind=dp) :: a_w_ice, pis, pvs
+    real(kind=dp) :: abifm_m, abifm_c
+    real(kind=dp) :: p_freeze, p_frozen
+
+    real(kind=dp) :: p_freeze_max, radius_max, diameter_max
+    logical :: freeze_thisTime
+
+    integer :: n_entry, ind, spec_bin = 13
+    integer :: k_th, n_parts_in_bin
+    real(kind=dp) :: aerosol_dry_radius, aerosol_radius, rand
+    real(kind=dp), allocatable :: H2O_masses(:), total_masses(:), &
+         H2O_frac(:)
+    integer :: i_spec_max
+    real(kind=dp) :: j_het_max
+    integer :: rand_geometric
+
+    integer :: loop_count = 0
+    
+
+    allocate(total_masses(aero_state_n_part(aero_state)))
+    allocate(H2O_masses(aero_state_n_part(aero_state)))
+    allocate(H2O_frac(aero_state_n_part(aero_state)))
```

Comment:
These don't seem to cause an error when removed. However maybe they could be kept to look similar to the others.